### ISSUE TITLE
Official Waterfox Current Linux build does not run

### ISSUE THREAD

**Issue by laniakea64** [BrowserWorks/waterfox#1128]
I use Xubuntu 18.04 64-bit, and I'm unable to run the official build of Waterfox 68.0b1 downloaded from the website.  It doesn't start at all.

If I try to run it in Terminal, I see this error -
```
XPCOMGlueLoad error for file /opt/waterfox/libmozsqlite3.so:
/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /opt/waterfox/libmozsqlite3.so)
Couldn't load XPCOM.
```

My custom build of 68.0b1 runs fine without error.

@MrAlex94 What distro/version are you using to build Waterfox for Linux?

**Comment by hawkeye116477** [comment]
Maybe problem can be solved by using statically linked libstdc++, just like for Classic.
Anyway, unofficial working package for 18.04 is on https://software.opensuse.org//download.html?project=home%3Ahawkeye116477%3Awaterfox&package=waterfox-beta (additionally it contains small fixes => `https://github.com/hawkeye116477/waterfox-deb/tree/master/waterfox-beta/patches`). 

There is also AppImage package with statically linked libstdc++ `https://download.opensuse.org/repositories/home:/hawkeye116477:/waterfox/AppImage/waterfox-beta-latest-x86_64.AppImage.mirrorlist`, which works even with CentOS 7.

Both packages are also compiled with LTO+PGO and gold as linker.

**Comment by grahamperrin** [comment]
> Xubuntu 18.04 64-bit,

OK on Lubuntu, touch wood: 

![image](https://user-images.githubusercontent.com/192271/64573876-e2fe0780-d364-11e9-89a3-57782f0150f1.png)

**Comment by laniakea64** [comment]
> OK on Lubuntu, touch wood:

@grahamperrin Could you please post the output from commands `uname -a` and `lsb_release -a` on that Lubuntu machine?

**Comment by grahamperrin** [comment]
18.10, compared to your (Kubuntu) 18.04

```text
grahamperrin@lubuntu:~$ uname -a
Linux lubuntu 4.18.0-25-generic #26-Ubuntu SMP Mon Jun 24 09:32:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
grahamperrin@lubuntu:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.10
Release:        18.10
Codename:       cosmic
grahamperrin@lubuntu:~$ 
```

**Comment by ilu33** [comment]
Same problem on Debian9:
```
dpkg -l libc6
<snip>
ii  libc6:amd64    2.24-11+deb9 amd64        GNU C Library: Shared libraries
ii  libc6:i386     2.24-11+deb9 i386         GNU C Library: Shared libraries
```
I am soon going to update to buster anyway. I'm just wondering whether this strict version requirement is really a good idea. Mozilla obviously doesn't think so.

**Comment by laniakea64** [comment]
In the case of Xubuntu 18.04, there is currently no OS upgrade option.  18.04 is the newest Ubuntu LTS.

And anyway, 18.04 is sufficiently new according to system requirements for Waterfox 68.0b1 listed on https://www.waterfox.net/releases/ -
> Linux
> Min. OS: Any modern distro

**Comment by hawkeye116477** [comment]
@laniakea64 There is 19.04 :smile:

**Comment by laniakea64** [comment]
> @laniakea64 There is 19.04 :smile:

19.04 is not LTS.

I can't afford the time to upgrade to an entire new OS version every 6 months.

**Comment by tcreek** [comment]
Just tested on Ubuntu `18.04, same issue

**Comment by reikred** [comment]
Is there a fix that will work for fedora 25? F25 has glibc version 2.24 and no upgrade available. Just tested also waterfox-current-2019.10, same problem

XPCOMGlueLoad error for file /usr/local/waterfox-current-2019.10/liblgpllibs.so:
/lib64/libm.so.6: version `GLIBC_2.27' not found (required by /usr/local/waterfox-current-2019.10/liblgpllibs.so)
Couldn't load XPCOM.

**Comment by tcreek** [comment]
@reikred You could try compiling the source with your glibc v2.24. I have noticed the beta builds have been removed. Updating glibc is not a trivial task as other applications are built on it, and using its dynamic libraries. All the others would have to be rebuilt on it.  

Building the source is really not that hard on Linux.

**Comment by hawkeye116477** [comment]
@reikred Fedora 25 is EOL, why you use still use that version?

**Comment by reikred** [comment]
@hawkeye I think if I could get a waterfox-current-2019.10 as an AppImage that would solve my problems on f25 for now :). UPDATE: I FOUND  Waterfox-2019.10-157.1.Build157.1.glibc2.17-x86_64.AppImage  and it runs ok !!

Background:  I was unfamiliar with AppImage, but have since I posted the above tried your AppImage version of 68.0b1, and that seems to work in f25. Yes, I know f25 is outdated and I am in the process of bringing up ubuntu 18.04 as a replacement but that is a slow process for unrelated reasons.

@tcreek, I will stick to pre-compiled if at all possible,. Thx, for the encouragement, though. I have had some "fun" compiling firefox several years ago ;).

**Comment by hawkeye116477** [comment]
`Waterfox-2019.10-157.1.Build157.1.glibc2.17-x86_64.AppImage and it runs ok !!`
Appears that I forgot to replace it with older version. That's little bad build, you can have some issues with extensions. It's also Classic, not Current. I'll publish new version later today.
I'm building AppImages with statically linked libstdc++ and using CentOS 7 packages, so that's why compatibility is better than in case of official package.

**Comment by reikred** [comment]
@hawkeye116477,

 I am watching https://download.opensuse.org/repositories/home:/hawkeye116477:/waterfox/AppImage/ for the new version. Apparently that is the URL at which the full list of available AppImage files is visible (mostly: note to self, but please correct me if I am looking in the wrong place :))

**Comment by hawkeye116477** [comment]
@reikred Latest fixed Classic is here now => https://download.opensuse.org/repositories/home:/hawkeye116477:/waterfox/AppImage/waterfox-classic-latest-x86_64.AppImage.mirrorlist. I changed name to stop confusion. Current will be later. Only problem in case of AppImages that automatic updates doesn't work and also it check for updates on `waterfox.net`, not on OBS, so little patch would be needed. I can't provide xml, so that's needs to be done somehow other, maybe with using libappimageupdate, but currently unfortunately something is broken on OBS and not only downloading updates is broken, but even checking for it doesn't work, it detects new version when it's not available.

**Comment by reikred** [comment]
@hawkeye116477. I'm only looking for AppImage for "Current". I already have 68.0a1 running in the official version. I will wait for waterfox-current-2019.10 AppImage to appear. Is there any estimate for when this might happen? Thx.

**Comment by hawkeye116477** [comment]
@reikred I'm sorry it took so long, but it's available now https://download.opensuse.org/repositories/home:/hawkeye116477:/waterfox/AppImage/waterfox-current-latest-x86_64.AppImage.mirrorlist with PGO+LTO and as a special bonus - language packs. But please be aware that you might experience bug with creating new profile every launch and to fix that you'll need every time run it from console with `--ProfileManager` or just unpack it. I don't know unfortunately how to fix that bug by another way, but seems that only happens only on some distros, for example on Manjaro works correct, only after few runs it can't start, but that can be fixed by removing  `addonStartup.json.lz4` from profile, maybe that's because I put unsigned langpacks into browser/extensions, but I set `extensions.autoDisableScopes` to `3`.

**Comment by reikred** [comment]
@hawkeye116477 I have it running . I was tripped up by the fact that wf-current switched back to ~/.waterfox as the profile location. whereas wf-68 is using ~/.mozilla/waterfox and wf-classic is using ~/.waterfox. Because of the  profile location change I have not been able to test extensively yet, and have not tested for the profile problems you mention.

There is also the separate issue https://github.com/MrAlex94/Waterfox/issues/1241 that makes me worried about running wf-current on existing profiles that works with other versions, so it may have to clone/rename some profiles for testing purposes, and it will take time.  But thanks for getting me something that runs !!!!

**Comment by grahamperrin** [comment]
> #1241 

Don't worry. The workaround is trivial, and one-off (once per downgrade): https://github.com/MrAlex94/Waterfox/issues/1241#issuecomment-547473005

> wf-68 

To avoid possible confusion: 

- Waterfox Current is (still) known as 68 in some contexts. This is by design.

`~/.mozilla/waterfox` arose with an early alpha of Waterfox Current.

**Comment by laniakea64** [comment]
Official Waterfox Current 2020.01 Linux build still does not run.  The error message has now changed -
```
XPCOMGlueLoad error for file /opt/waterfox/libnspr4.so:
/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.30' not found (required by /opt/waterfox/libnspr4.so)
Couldn't load XPCOM.
```

> @MrAlex94 What distro/version are you using to build Waterfox for Linux?

**Comment by InconsolableCellist** [comment]
Doesn't work here either, with waterfox-classic on Ubuntu 18.10

```
$ uname -a 
Linux ... 4.18.0-13-generic #14-Ubuntu SMP Wed Dec 5 09:04:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.10 (Cosmic Cuttlefish)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.10"
VERSION_ID="18.10"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=cosmic
UBUNTU_CODENAME=cosmic

$ ./waterfox -P
XPCOMGlueLoad error for file /home/user/Downloads/waterfox-classic/libnspr4.so:
/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.30' not found (required by /home/user/Downloads/waterfox-classic/libnspr4.so)
Couldn't load XPCOM.

$ ldd --version
ldd (Ubuntu GLIBC 2.28-0ubuntu1) 2.28
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

Extracted from: waterfox-classic-2020.01.en-US.linux-x86_64.tar.bz2

**Comment by ilu33** [comment]
I can confirm that Waterfox current 2020.01 does not work on debian 9 and debian 10.
```
/opt/waterfox-current-test/waterfox
XPCOMGlueLoad error for file /opt/waterfox-current-test/libnspr4.so:
/lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.30' not found (required by /opt/waterfox-current-test/libnspr4.so)
Couldn't load XPCOM.

ldd --version
ldd (Debian GLIBC 2.28-10) 2.28
Copyright (C) 2018 Free Software Foundation, Inc. <snip>

**Comment by laniakea64** [comment]
https://www.waterfox.net/download/ now says
> Linux
> Minimum Requirements
> 
>    GLib 2.28

However, the official Waterfox Current 2020.01.1 Linux build doesn't run even in CentOS 8 -
```
XPCOMGlueLoad error for file /opt/waterfox/libmozsandbox.so:
/lib64/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /opt/waterfox/libmozsandbox.so)
Couldn't load XPCOM.
```

**Comment by MrAlex94** [comment]
So I went back to Ubuntu 16 LTS to compile Waterfox for 2020.01.1. This is so frustrating, the dev experience is not optimal on Linux, can't believe it's so complicated to have a portable application.

I'll see what I can do for the next build, but it's taking up so much time.

**Comment by jcea** [comment]
> So I went back to Ubuntu 16 LTS to compile Waterfox for 2020.01.1. This is so frustrating, the dev experience is not optimal on Linux, can't believe it's so complicated to have a portable application.
> 
> I'll see what I can do for the next build, but it's taking up so much time.

May I suggest a virtual machine to compile the release build? Ideally, it should be simply "git checkout", "make build", "make publish".

You don't need to develop Waterfox in Ubuntu 16.04, only compile there the release build.

Let me know if you need help to set this up.

**Comment by laniakea64** [comment]
> So I went back to Ubuntu 16 LTS to compile Waterfox for 2020.01.1. 

This solved this problem for Classic 2020.01.1.  But did not work for Current 2020.01.1? :confused: 

Are you using the same build machine for both Waterfoxes?

**Comment by ilu33** [comment]
On debian stable it does not work for classic see https://github.com/MrAlex94/Waterfox/issues/1356#issuecomment-585808751 It seems that Ubuntu 16.04 is not conservative enough (no surprise really).

I did not test current. Edit: Same with current

**Comment by laniakea64** [comment]
Official Waterfox Current 2020.02 Linux build works for me in Xubuntu 18.04 after running `sudo apt-get install libc++1` :tada: 
Same for Debian 10.3 live CD.

It also works in Xubuntu 16.04 after also installing `libc++abi1`.

**Comment by ilu33** [comment]
libc++1-7 and libc++abi1-7 were already installed. additionally installing libc++1 and libc++abi1 did not help, these packages do not contain anything (see https://packages.debian.org/buster/amd64/libc++1/filelist and https://packages.debian.org/buster/amd64/libc++abi1/filelist). 

There must be something else. Which debian live CD did you use?

**Comment by laniakea64** [comment]
> Which debian live CD did you use?

Debian 10.3.0 amd64 XFCE live CD from https://www.debian.org/CD/live/

**Comment by ViniciusBRodrigues** [comment]
I'm having the same issue on the linux mint debbie.

```
 ./waterfox
XPCOMGlueLoad error for file /home/viniciusbr/Downloads/waterfox/libnspr4.so:
/lib/x86_64-linux-gnu/libc.so.6: version GLIBC_2.30 not found (required by /home/viniciusbr/Downloads/waterfox/libnspr4.so)
Couldn't load XPCOM.
```

```
sudo apt install libnspr4
libnspr4 is already the newest version (2:4.20-1).
```

```
sudo apt install libc6
libc6 is already the newest version (2.28-10).
```

**Comment by cocoa1231** [comment]
Same issue here on LMDE

  ```
$ ./waterfox   
  XPCOMGlueLoad error for file /home/cocoa/Applications/waterfox/libnspr4.so:
  /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.30' not found (required by /home/cocoa/Applications/waterfox/libnspr4.so)
  Couldn't load XPCOM.

```

**Comment by tcreek** [comment]
> I'm having the same issue on the linux mint debbie.
> 
> ```
>  ./waterfox
> XPCOMGlueLoad error for file /home/viniciusbr/Downloads/waterfox/libnspr4.so:
> /lib/x86_64-linux-gnu/libc.so.6: version GLIBC_2.30 not found (required by /home/viniciusbr/Downloads/waterfox/libnspr4.so)
> Couldn't load XPCOM.
> ```
> 
> ```
> sudo apt install libnspr4
> libnspr4 is already the newest version (2:4.20-1).
> ```
> 
> ```
> sudo apt install libc6
> libc6 is already the newest version (2.28-10).
> ```

You seem to be misreading the output it is giving you

: version **GLIBC_2.30** not found 

libc6 is already the newest version **(2.28-10).**

Since I doubt your skill level is sufficient to put version 2.30+ on your system, you are going to have to upgrade to a newer distro

**Comment by AngborTim** [comment]
The same issue

Linux Big 5.4.0-72-lowlatency #80-Ubuntu SMP PREEMPT Mon Apr 12 18:37:24 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
ldd (Ubuntu GLIBC 2.31-0ubuntu9.3) 2.31
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

./waterfox
Illegal instruction

**Comment by hawkeye116477** [comment]
@AngborTim From your log, seems that's rather different problem. Official Waterfox G4 requires SSE4.2 instruction support in CPU and seems that your CPU doesn't support that. However there is solution for you other than buying new CPU => just use one of my packages following https://github.com/hawkeye116477/waterfox-deb-rpm-arch-AppImage/blob/master/README.md.

**Comment by AngborTim** [comment]
> @AngborTim From your log, seems that's rather different problem. Official Waterfox G4 requires SSE4.2 instruction support in CPU and seems that your CPU doesn't support that. However there is solution for you other than buying new CPU => just use one of my packages following https://github.com/hawkeye116477/waterfox-deb-rpm-arch-AppImage/blob/master/README.md.

thanx, I'll check it
